In [1]:
import pandas as pd

In [2]:
df=pd.read_excel(r"D:\Tableau Files\Practice files course 7\DS1_C7_S4_Project_BirdStrike_Data.xlsx")

In [3]:
df.shape

(25558, 26)

In [4]:
# printing basic info and null records, duplicates about dataset

print(df.info())
print("-----------")
print("Missing values")
print(df.isnull().sum())
print("-----------")
print("Duplicates:", df.duplicated().sum())
print("-----------")
print(df.nunique())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25558 entries, 0 to 25557
Data columns (total 26 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   Record ID                                25558 non-null  int64         
 1   Aircraft: Type                           25429 non-null  object        
 2   Airport: Name                            25429 non-null  object        
 3   Altitude bin                             25429 non-null  object        
 4   Aircraft: Make/Model                     25558 non-null  object        
 5   Wildlife: Number struck                  25429 non-null  object        
 6   Wildlife: Number Struck Actual           25558 non-null  int64         
 7   Effect: Impact to flight                 2078 non-null   object        
 8   FlightDate                               25429 non-null  datetime64[ns]
 9   Effect: Indicated Damage               

In [5]:
for column in df.columns:
    print(f"Column: {column}")
    print(df[column].value_counts())
    print("\n")

Column: Record ID
Record ID
202152    1
254643    1
258806    1
257552    1
257811    1
         ..
218134    1
215563    1
220925    1
218412    1
319593    1
Name: count, Length: 25558, dtype: int64


Column: Aircraft: Type
Aircraft: Type
Airplane    25429
Name: count, dtype: int64


Column: Airport: Name
Airport: Name
DALLAS/FORT WORTH INTL ARPT    803
SACRAMENTO INTL                676
SALT LAKE CITY INTL            479
DENVER INTL AIRPORT            476
KANSAS CITY INTL               452
                              ... 
ORCAS ISLAND ARPT                1
T B LETTSOME INTL                1
DYERSBURG REGIONAL ARPT          1
MCCORD FIELD ARPT                1
PRINCESS JULIANA                 1
Name: count, Length: 1109, dtype: int64


Column: Altitude bin
Altitude bin
< 1000 ft    20556
> 1000 ft     4873
Name: count, dtype: int64


Column: Aircraft: Make/Model
Aircraft: Make/Model
B-737-700        2488
B-737-300        2309
CL-RJ100/200     1951
A-320            1193
A-319       

In [6]:
# Replacing None
df['Effect: Impact to flight'] = df['Effect: Impact to flight'].fillna('No Effect')
df['Conditions: Precipitation'] = df['Conditions: Precipitation'].fillna('No Precipitation')

# Replacing 'Y': 'Yes', 'N': 'No'
df['Pilot warned of birds or wildlife?'] = df['Pilot warned of birds or wildlife?'].replace({'Y': 'Yes', 'N': 'No'})

# Fill categorical missing values with mode
categorical_cols = ["Aircraft: Type", "Airport: Name", "Altitude bin", 
                    "Aircraft: Airline/Operator", "When: Phase of flight", "Is Aircraft Large?"]
for col in categorical_cols:
    df[col] = df[col].fillna(df[col].mode()[0])  # Use assignment instead of inplace

# Fill with 'Unknown' for certain categorical columns
unknown_fill = ["Wildlife: Number struck", "Wildlife: Size", "Pilot warned of birds or wildlife?", "Origin State"]
df[unknown_fill] = df[unknown_fill].fillna("Unknown")  # Already correct

# Fill numerical missing values
df["Aircraft: Number of engines?"] = df["Aircraft: Number of engines?"].fillna(df["Aircraft: Number of engines?"].mode()[0])  # Use assignment
df["Feet above ground"] = df["Feet above ground"].fillna(df["Feet above ground"].median())  # Use assignment
df["FlightDate"] = df["FlightDate"].ffill()

# Fill 'Remarks' column with "No remarks"
df["Remarks"] = df["Remarks"].fillna("No remarks")  # Use assignment

### **Data Preprocessing & Observations** (For PPT)

#### **Data Cleaning**

*   Replaced None with No effect in columns: `"Effect: Impact to flight"`, `"Conditions: Precipitation"`
*   Filled categorical values using **mode** or **"Unknown"**
*   Filled numerical values using **median/mode**
*   `"Remarks"` → Replaced missing with **"No remarks"**
*   **Forward-filled missing `FlightDate`** → Preserved chronological consistency.
*   No **duplicates**; data is **consistent & reliable**

#### **Key Observations**

*   **25,558 wildlife strikes** recorded
*   `"Wildlife: Species"` complete → analyze species impact
*   `"Phase of flight"` helps identify **high-risk moments**
*   `"Cost: Total $"` enables **financial impact analysis**
*   `"Number of people injured"` aids in **safety assessment**

In [8]:
# Mapping dictionary for states/provinces to countries
state_to_country = {
    # U.S. States
    'California': 'United States',
    'Texas': 'United States',
    'Florida': 'United States',
    'New York': 'United States',
    'Illinois': 'United States',
    'Pennsylvania': 'United States',
    'Missouri': 'United States',
    'Kentucky': 'United States',
    'Ohio': 'United States',
    'Hawaii': 'United States',
    'Michigan': 'United States',
    'Colorado': 'United States',
    'Tennessee': 'United States',
    'North Carolina': 'United States',
    'New Jersey': 'United States',
    'Georgia': 'United States',
    'Utah': 'United States',
    'Maryland': 'United States',
    'DC': 'United States',
    'Louisiana': 'United States',
    'Oregon': 'United States',
    'Nebraska': 'United States',
    'Arizona': 'United States',
    'Massachusetts': 'United States',
    'Minnesota': 'United States',
    'Indiana': 'United States',
    'Washington': 'United States',
    'Alabama': 'United States',
    'Virginia': 'United States',
    'Wisconsin': 'United States',
    'Connecticut': 'United States',
    'Iowa': 'United States',
    'Oklahoma': 'United States',
    'Alaska': 'United States',
    'Nevada': 'United States',
    'South Carolina': 'United States',
    'Rhode Island': 'United States',
    'New Hampshire': 'United States',
    'Arkansas': 'United States',
    'Mississippi': 'United States',
    'Kansas': 'United States',
    'New Mexico': 'United States',
    'North Dakota': 'United States',
    'West Virginia': 'United States',
    'Puerto Rico': 'United States',
    'Idaho': 'United States',
    'South Dakota': 'United States',
    'Montana': 'United States',
    'Maine': 'United States',
    'Vermont': 'United States',
    'Wyoming': 'United States',
    'Virgin Islands': 'United States',
    'Delaware': 'United States',
    'Prince Edward Island': 'Canada',  # Canadian Provinces
    'Ontario': 'Canada',
    'British Columbia': 'Canada',
    'Quebec': 'Canada',
    'Alberta': 'Canada',
    'Newfoundland and Labrador': 'Canada',
    'Saskatchewan': 'Canada'}

# Add the 'Country' column using the mapping dictionary
df['Country'] = df['Origin State'].map(state_to_country)

# Fill missing values (if any) with 'Unknown'
df['Country'] = df['Country'].fillna('Unknown')

In [9]:
# Save cleaned data
df.to_excel(r"D:\Tableau Files\Practice files course 7\c7 project\Bird_Strike_Data.xlsx", index=False)

print("Missing values handled successfully!")

print("Missing values")
print(df.isnull().sum())

Missing values handled successfully!
Missing values
Record ID                                  0
Aircraft: Type                             0
Airport: Name                              0
Altitude bin                               0
Aircraft: Make/Model                       0
Wildlife: Number struck                    0
Wildlife: Number Struck Actual             0
Effect: Impact to flight                   0
FlightDate                                 0
Effect: Indicated Damage                   0
Aircraft: Number of engines?               0
Aircraft: Airline/Operator                 0
Origin State                               0
When: Phase of flight                      0
Conditions: Precipitation                  0
Remains of wildlife collected?             0
Remains of wildlife sent to Smithsonian    0
Remarks                                    0
Wildlife: Size                             0
Conditions: Sky                            0
Wildlife: Species                          0
Pil

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25558 entries, 0 to 25557
Data columns (total 27 columns):
 #   Column                                   Non-Null Count  Dtype         
---  ------                                   --------------  -----         
 0   Record ID                                25558 non-null  int64         
 1   Aircraft: Type                           25558 non-null  object        
 2   Airport: Name                            25558 non-null  object        
 3   Altitude bin                             25558 non-null  object        
 4   Aircraft: Make/Model                     25558 non-null  object        
 5   Wildlife: Number struck                  25558 non-null  object        
 6   Wildlife: Number Struck Actual           25558 non-null  int64         
 7   Effect: Impact to flight                 25558 non-null  object        
 8   FlightDate                               25558 non-null  datetime64[ns]
 9   Effect: Indicated Damage               